File: mobile/src/main/java/com/keepnotes/models/Note.java
package com.keepnotes.models;

import android.os.Parcel;
import android.os.Parcelable;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.List;

public class Note implements Parcelable {
    private long id;
    private String title;
    private String description;
    private boolean isDone;

    public Note() {
    }

    public Note(long id, String title, String description, boolean isDone) {
        this.id = id;
        this.title = title;
        this.description = description;
        this.isDone = isDone;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public boolean isDone() {
        return isDone;
    }

    public void setDone(boolean done) {
        isDone = done;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(this.id);
        dest.writeString(this.title);
        dest.writeString(this.description);
        dest.writeByte(isDone ? (byte) 1 : (byte) 0);
    }

    protected Note(Parcel in) {
        this.id = in.readLong();
        this.title = in.readString();
        this.description = in.readString();
        this.isDone = in.readByte() != 0;
    }

    public static final Creator<Note> CREATOR = new Creator<Note>() {
        public Note createFromParcel(Parcel source) {
            return new Note(source);
        }

        public Note[] newArray(int size) {
            return new Note[size];
        }
    };

    public static Type getListTokenType() {
        return new TypeToken<List<Note>>() {
        }.getType();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Note note = (Note) o;

        return id == note.id;

    }

    @Override
    public int hashCode() {
        return (int) (id ^ (id >>> 32));
    }
}

File: mobile/src/main/java/com/keepnotes/models/Note.kt
package com.keepnotes.models

import android.os.Parcel
import android.os.Parcelable
import com.google.gson.reflect.TypeToken

import java.lang.reflect.Type

class Note : Parcelable {
    var id: Long = 0
    var title: String = ""
    var description: String? = null
    var isDone: Boolean = false

    constructor() {}

    constructor(id: Long, title: String, description: String, isDone: Boolean) {
        this.id = id
        this.title = title
        this.description = description
        this.isDone = isDone
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeLong(this.id)
        dest.writeString(this.title)
        dest.writeString(this.description)
        dest.writeByte(if (isDone) 1.toByte() else 0.toByte())
    }

    private constructor(`in`: Parcel) {
        this.id = `in`.readLong()
        this.title = `in`.readString()
        this.description = `in`.readString()
        this.isDone = `in`.readByte().toInt() != 0
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false

        val note = o as Note?

        return id == note!!.id

    }

    override fun hashCode(): Int {
        return (id xor id.ushr(32)).toInt()
    }

    companion object {

        val CREATOR: Parcelable.Creator<Note> = object : Parcelable.Creator<Note> {
            override fun createFromParcel(source: Parcel): Note {
                return Note(source)
            }

            override fun newArray(size: Int): Array<Note?> {
                return arrayOfNulls(size)
            }
        }

        val listTokenType: Type
            get() = object : TypeToken<List<Note>>() {

            }.type
    }
}
